src/Language/DifferentialDatalog/TypeInference.hs

```diff
@@ -0,0 +1,637 @@
+             -- unknown type arguments of function calls.  We could use
+             -- arbitrary auto-generated names for them, but to avoid a state
+             -- monad, we use context and type argument name to uniquely
+             -- identify the variable.
+           | TETVar ECtx String
+
+teTuple :: [TExpr] -> TExpr
+teTuple [t] = t
+teTuple ts  = TETuple (Just $ length ts) (M.fromList $ mapIdx (\t i -> (i, t)) ts)
+
+-- Type congruence, e.g., '|e| = BigInt'.
+data Predicate = PEq TExpr TExpr
+
+(===) :: TExpr -> TExpr -> Constraint
+(===) t1 t2 = CPredicate $ PEq t1 t2
+
+-- Type expression is a struct of the specified user-defined type: 'is_MyStruct |e|'.
+deIsStruct :: (?d::DatalogProgram) => DDExpr -> String -> Constraint
+deIsStruct de n = CPredicate $ deIsStruct_ de n
+
+deIsStruct_ :: (?d::DatalogProgram) => DDExpr -> String -> Predicate
+deIsStruct_ de@(DDExpr ctx _) n =
+    PEq (TETypeOfExpr de) (TEUser n (map (\a -> TETVar ctx $ name a) tdefArgs))
+    where
+    TypeDef{..} = getType ?d n
+
+deIsBool :: DDExpr -> Constraint
+deIsBool de = TETypeOfExpr de === TEBool
+
+deIsString :: DDExpr -> Constraint
+deIsString de = CPredicate $ deIsString_ de
+
+deIsString_ :: DDExpr -> Predicate
+deIsString_ de = PEq (TETypeOfExpr de) TEString
+
+deIsBit :: DDExpr -> Constraint
+deIsBit de = CPredicate $ deIsBit_ de
+
+deIsBit_ :: DDExpr -> Predicate
+deIsBit_ de = PEq (TETypeOfExpr de) (TEBit (IWidthOfExpr de))
+
+deIsSigned_ :: DDExpr -> Predicate
+deIsSigned_ de = PEq (TETypeOfExpr de) (TESigned (IWidthOfExpr de))
+
+deIsBigInt_ :: DDExpr -> Predicate
+deIsBigInt_ de = PEq (TETypeOfExpr de) TEBigInt
+
+deIsFloat :: DDExpr -> Constraint
+deIsFloat de = CPredicate $ deIsFloat_ de
+
+deIsFloat_ :: DDExpr -> Predicate
+deIsFloat_ de = PEq (TETypeOfExpr de) TEFloat
+
+deIsDouble_ :: DDExpr -> Predicate
+deIsDouble_ de = PEq (TETypeOfExpr de) TEDouble
+
+deIsFP :: DDExpr -> Constraint
+deIsFP de = CDisj $ deIsFP_ de
+
+deIsFP_ :: DDExpr -> [[Predicate]]
+deIsFP_ de = [[deIsFloat_ de], [deIsDouble_ de]]
+
+-- 'is_bits t = is_Bit t || is_Signed t'.
+deIsBits :: DDExpr -> Constraint
+deIsBits de = CDisj $ deIsBits_ de
+
+deIsBits_ :: DDExpr -> [[Predicate]]
+deIsBits_ de = [[deIsBit_ de], [deIsSigned_ de]]
+
+-- 'is_int t = is_bits t || is_BigInt t'.
+deIsInt :: DDExpr -> Constraint
+deIsInt de = CDisj $ deIsInt_ de
+
+deIsInt_ :: DDExpr -> [[Predicate]]
+deIsInt_ de = deIsBits_ de ++ [[deIsBigInt_ de]]
+
+-- 'is_num t = is_int t || is_fp t'.
+deIsNum :: DDExpr -> Constraint
+deIsNum de = CDisj $ deIsInt_ de ++ deIsFP_ de
+
+-- Type expression is a shared reference type with specified inner type.
+deIsSharedRef :: (?d::DatalogProgram) => DDExpr -> TExpr -> [Constraint]
+deIsSharedRef de t =
+    cdisj $ map (\tdef -> [PEq (TETypeOfExpr de) (TEExtern (name tdef) [t])]) sref_tdefs
+    where
+    -- All shared reference types.
+    sref_tdefs = filter (tdefGetSharedRefAttr ?d)
```

Comment:
I find it a bit strange that you can lookup stuff in the program while creating constraints, but I don't see why this would not work.